pkg/ingester/instance.go

```diff
@@ -381,27 +381,48 @@ func (i *instance) QuerySample(ctx context.Context, req logql.SelectSampleParams
 	return iters, nil
 }
 
-func (i *instance) Label(_ context.Context, req *logproto.LabelRequest) (*logproto.LabelResponse, error) {
-	var labels []string
-	if req.Values {
-		values, err := i.index.LabelValues(req.Name, nil)
-		if err != nil {
-			return nil, err
-		}
-		labels = make([]string, len(values))
-		for i := 0; i < len(values); i++ {
-			labels[i] = values[i]
+func (i *instance) Label(_ context.Context, req *logproto.LabelRequest, matchers ...*labels.Matcher) (*logproto.LabelResponse, error) {
+	if len(matchers) == 0 {
+		var labels []string
+		if req.Values {
+			values, err := i.index.LabelValues(req.Name, nil)
+			if err != nil {
+				return nil, err
+			}
+			labels = make([]string, len(values))
+			for i := 0; i < len(values); i++ {
+				labels[i] = values[i]
+			}
+		} else {
+			names, err := i.index.LabelNames(nil)
+			if err != nil {
+				return nil, err
+			}
+			labels = make([]string, len(names))
+			for i := 0; i < len(names); i++ {
+				labels[i] = names[i]
+			}
 		}
-	} else {
-		names, err := i.index.LabelNames(nil)
-		if err != nil {
-			return nil, err
+		return &logproto.LabelResponse{
+			Values: labels,
+		}, nil
+	}
+
+	ids, _ := i.index.Lookup(matchers, nil)
+	labels := make([]string, 0)
+	for _, streamID := range ids {
+		stream, ok := i.streamsByFP[streamID]
+		if !ok {
+			return nil, ErrStreamMissing
 		}
-		labels = make([]string, len(names))
-		for i := 0; i < len(names); i++ {
-			labels[i] = names[i]
+
+		for _, label := range stream.labels {
+			if label.Name == req.Name {
```

Comment:
What if `req.Name` is nil, in the case of [`api/v1/labels`](https://grafana.com/docs/loki/latest/api/#get-lokiapiv1labels)?

Reply:
A separate branch for getting the label names has been added.